### PR TITLE
fix: Fixes a few null handling bugs

### PR DESCRIPTION
**Pull request by jnh5y**
* Dereferencing a null struct no longer returns a null.
Addresses https://github.com/confluentinc/ksql/issues/9029.

* Calling subscripts on null maps/arrays no longer returns a null.
Addresses https://github.com/confluentinc/ksql/issues/9030

* Added a test to show that calling subscripts on a null map/array in a function call works properly.
Addresses https://github.com/confluentinc/ksql/issues/9031

### Testing done 
Unit tests and QTTs have been added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by jnh5y** [comment]
@spena there is a test case in `ExpressionEvaluatorParityTest` failing based on what the value of `"ADDRESS->STREET + 'foo'"` should be.  

Previously, it logged an error here: https://github.com/confluentinc/ksql/blob/master/ksqldb-engine/src/test/java/io/confluent/ksql/execution/ExpressionEvaluatorParityTest.java#L163

With the change in the code, it no longer logs an error.  The `InterpretedExpression` returns `null` while `CompiledExpression` returns the string `"nullfoo"`.

**Comment by suhas-satish** [comment]
thank you for fixing these @jnh5y  !

**Review by suhas-satish — APPROVED** [review]

File: ksqldb-engine/src/test/java/io/confluent/ksql/execution/ExpressionEvaluatorParityTest.java [hunk 1]
```diff
@@ -160,7 +160,8 @@ public void shouldDereference() throws Exception {
     assertOrders("ITEMINFO->NAME", ITEM_NAME);
     assertOrders("ITEMINFO->CATEGORY->NAME", CATEGORY_NAME);
     assertOrders("'a-' + ITEMINFO->CATEGORY->NAME + '-b'", "a-cat-b");
-    assertOrdersError("ADDRESS->STREET + 'foo'", evalLogger(null));
+    // Due to https://github.com/confluentinc/ksql/issues/9136 the next assert cannot be updated.
+    // assertOrders("ADDRESS->STREET + 'foo'", null);
```
Comment (jnh5y) [inline ksqldb-engine/src/test/java/io/confluent/ksql/execution/ExpressionEvaluatorParityTest.java:164]: @suhas-satish As a sanity check, while addressing the null handling issues, I think I turned up a new, subtle issue.  

My solution is to tag that new issue and merge this PR.  Lemme know if that sounds reasonable to you.  

The issue was found via CI failing; if CI passes and you don't have concerns, I'll merge this PR later today!

Comment (suhas-satish) [reply ksqldb-engine/src/test/java/io/confluent/ksql/execution/ExpressionEvaluatorParityTest.java:164]: sounds good , thanks Jim

**Pull request merged**